src/cargo-fmt/main.rs

```diff
@@ -446,10 +537,99 @@ fn get_targets_with_hitlist(
     }
 }
 
-fn add_targets(target_paths: &[cargo_metadata::Target], targets: &mut BTreeSet<Target>) {
+fn add_targets(
+    manifest_path: &PathBuf,
+    target_paths: &[cargo_metadata::Target],
+    targets: &mut BTreeSet<Target>,
+    include_nested_test_files: bool,
+    ignore_list: &Vec<String>,
+) {
+    let mut test_files_added = false;
     for target in target_paths {
-        targets.insert(Target::from_target(target));
+        // Packages often have more than one `test` target,
+        // so only add the nested files for the first one.
+        let test_files = if include_nested_test_files
+            && !test_files_added
+            && target.kind.iter().any(|t| t == "test")
+        {
+            // If any errors are encountered, just fallback to ignoring
+            // nested files in subdirectories of `tests`
+            match manifest_path.parent() {
+                None => None,
+                Some(package_dir) => {
+                    let target_dir = package_dir.join("tests");
+                    let mut ignore_builder = gitignore::GitignoreBuilder::new(package_dir);
+                    let mut had_ignore_add_error = false;
+
+                    for ignore_path in ignore_list {
+                        if ignore_builder.add_line(None, &ignore_path).is_err() {
+                            had_ignore_add_error = true;
+                            break;
+                        };
+                    }
+
+                    if had_ignore_add_error {
+                        None
+                    } else {
+                        match ignore_builder.build() {
+                            Err(_) => None,
+                            Ok(ignore_set) => {
+                                test_files_added = true;
+                                Some(get_nested_integration_test_files(
+                                    &target_dir,
+                                    &target_dir,
+                                    &ignore_set,
+                                ))
+                            }
+                        }
+                    }
+                }
+            }
+        } else {
+            None
+        };
+        targets.insert(Target::from_target(target, test_files));
+    }
+}
+
+// Returns a `Vec` containing `PathBuf`s of files nested .rs files within a subdirectory
+// under the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(
+    path: &Path,
+    root_dir: &Path,
+    ignore_set: &gitignore::Gitignore,
+) -> Vec<PathBuf> {
+    let mut files = vec![];
+    if path.is_dir()
+        && !ignore_set
+            .matched_path_or_any_parents(&path, true)
+            .is_ignore()
+    {
+        for entry in fs::read_dir(path).expect(&format!(
+            "couldn't read directory {}",
+            path.to_str().unwrap()
```

Comment:
This function was pretty much a copy/pasta from [here](https://github.com/rust-lang/rustfmt/blob/master/src/test/mod.rs#L91). I left the `.expect(`'s in there for now. Can clean them up if folks think the overall approach in this PR might be viable